libbpf-tools/execsnoop.c

```diff
@@ -0,0 +1,340 @@
+	{ "print-uid", 'U', NULL, 0, "print UID column"},
+	{ "max-args", MAX_ARGS_KEY, "MAX_ARGS", 0,
+		"maximum number of arguments parsed and displayed, defaults to 20"},
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	long int uid, max_args;
+
+	switch (key) {
+	case 'h':
+		argp_usage(state);
+		break;
+	case 'T':
+		env.time = true;
+		break;
+	case 't':
+		env.timestamp = true;
+		break;
+	case 'x':
+		env.fails = true;
+		break;
+	case 'u':
+		errno = 0;
+		uid = strtol(arg, NULL, 10);
+		if (errno || uid < 0 || uid >= INVALID_UID) {
+			fprintf(stderr, "Invalid UID %s\n", arg);
+			argp_usage(state);
+		}
+		env.uid = uid;
+		break;
+	case 'q':
+		env.quote = true;
+		break;
+	case 'n':
+		env.name = arg;
+		break;
+	case 'l':
+		env.line = arg;
+		break;
+	case 'U':
+		env.print_uid = true;
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case MAX_ARGS_KEY:
+		errno = 0;
+		max_args = strtol(arg, NULL, 10);
+		if (errno || max_args < 1 || max_args > TOTAL_MAX_ARGS) {
+			fprintf(stderr, "Invalid MAX_ARGS %s, should be in [1, %d] range\n",
+					arg, TOTAL_MAX_ARGS);
+
+			argp_usage(state);
+		}
+		env.max_args = max_args;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+            const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void time_since_start()
+{
+	long nsec, sec;
+	static struct timespec cur_time;
+	double time_diff;
+
+	clock_gettime(CLOCK_MONOTONIC, &cur_time);
+	nsec = cur_time.tv_nsec - start_time.tv_nsec;
+	sec = cur_time.tv_sec - start_time.tv_sec;
+	if (nsec < 0) {
+		nsec += NSEC_PER_SEC;
+		sec--;
+	}
+	time_diff = sec + (double)nsec / NSEC_PER_SEC;
+	printf("%-8.3f", time_diff);
+}
+
+static void inline quoted_symbol(char c) {
+	switch(c) {
+		case '"':
+			putchar('\\');
+			putchar('"');
+			break;
+		case '\t':
+			putchar('\\');
+			putchar('t');
+			break;
+		case '\n':
+			putchar('\\');
+			putchar('n');
+			break;
+		default:
+			putchar(c);
+			break;
+	}
+}
+
+static void print_args(const struct event *e, bool quote)
+{
+	int args_counter = 0;
+
+	if (env.quote)
+		putchar('"');
+
+	for (int i = 0; i < e->args_size && args_counter < e->args_count; i++) {
+		char c = e->args[i];
+		if (env.quote) {
+			if (c == '\0') {
+				args_counter++;
+				putchar('"');
+				putchar(' ');
+				if (args_counter < e->args_count) {
+					putchar('"');
+				}
+			} else {
+				quoted_symbol(c);
+			}
+		} else {
+			if (c == '\0') {
+				args_counter++;
+				putchar(' ');
+			} else {
+				putchar(c);
+			}
+		}
+	}
+	if (e->args_count == env.max_args) {
```

Comment:
let me ask again...

> As for args_count == max_args wouldn't it be the case when actual number of args is exactly equal to max_args?

Am I missing why `args_count == max_args` is sufficient to deduce " ..." part?

Reply:
It is the same behavior as `more_args` after loop and same as original [execsnoop](https://github.com/iovisor/bcc/blob/master/tools/execsnoop.py#L181) implementation, and same behavior as [your sample](https://github.com/anakryiko/bcc/commit/cda6c3411787504bc7c1dcf8af419a6e86b923c3#diff-ee7df97255c38d4d9f9fa421b1922f9eR154).

Original execsnoop doesn't distinguish results from `args > max_args` and `args >= max_args`. Here's an example:

   #script1
   /bin/echo 1 2 3 4
   /bin/echo 1 2 3

   $ ./execsnoop.py --max-args=4
    PCOMM            PID    PPID   RET ARGS
    echo             7730   7487     0 /bin/echo 1 2 3 ...
    echo             7731   7487     0 /bin/echo 1 2 3 ...

If another behavior is intended (that distinguish between `args == max_args` and `args > max_args`) in that case implementation should read one argument more (beyond `max_args`). But for such argument implementation could use much smaller read size than `ARGSIZE`.

Reply:
Also, I want to mention that this also applies for argument size too, cause currently all execsnoop implementation doesn't indicated that argument was truncated and the only indication of such possibility is argument that has length of `ARGSIZE`.

Reply:
Right, so existing execsnoop has a bug, and the answer is to preserve it instead of fixing it? Especially given that implementation is very simple (btw, NULL initialization of argp is not necessary):
```
for (int i = 1; i <= TOTAL_MAX_ARGS; i++) {
        /* NULL initialization is not needed */
		const char *argp;

		bpf_probe_read(&argp, sizeof(argp), &args[i]);
		if (!argp)
			return 0;

		if (i == max_args) {
		    event->args_count = -event->args_count;
		    /* or whatever way you insist on */
		    return 0;
		}

		if (event->args_size > LAST_ARG)
		    return 0;

		ret = bpf_probe_read_str(&event->args[event->args_size], ARGSIZE, argp);
		if (ret > ARGSIZE)
			return 0;

		event->args_count++;
		event->args_size += ret;
}
```

My demo already handled such case, btw, if you didn't like that, I'd appreciate you mentioning that, at least, instead of me having to do few more rounds of comments to get you to reply to my concern.

And yes, argument truncation is not indicated in any way, and I think it would be nice if both execsnoop variants would do that. But that can be done separately from this PR.

Reply:
Your demo have exactly same behavior (I tested before writing), and this behavior isn't a bug.

Reply:
Here's an output from your sample:
```
[15826/15826] bash (ret=0), filename /bin/echo, 30 args:/bin/echo,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,2
6,27,28,29
[15827/15827] bash (ret=0), filename /bin/echo, 30 args:/bin/echo,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,2
6,27,28,29
```
For: 
```
/bin/echo 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29
$ /bin/echo 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30
```

As I [mentioned](https://github.com/iovisor/bcc/pull/2807#discussion_r425928238) that same behavior was in [your sample](https://github.com/anakryiko/bcc/commit/cda6c3411787504bc7c1dcf8af419a6e86b923c3#diff-ee7df97255c38d4d9f9fa421b1922f9eR154), moreover you did exactly same comparison (only in bpf code) `if (i == MAX_ARG_CNT)` but for some reason if it succeeded you negated it's value (which could be done in userspace and would achieve same result).

Reply:
Moreover, your sample had extra bug (when `MAX_ARG_CNT` isn't equal to `max_args`).

Reply:
```
if (i == max_args) {
	event->args_count = -event->args_count;
	/* or whatever way you insist on */
	return 0;
}
```
Again, this code has same **bug**, cause it just checks that `i == max_args` which could be done in userspace and would have same result. As I wrote before:

> in that case implementation should **read one argument more** (**beyond** max_args). But for such argument implementation could use **much smaller read size** than ARGSIZE.

>  (btw, NULL initialization of argp is not necessary)

Can you explain why ?

Reply:
> Your demo have exactly same behavior (I tested before writing), and this behavior isn't a bug.

Yes, good catch, my demo did have those two bugs, but its intent was to handle this correctly (otherwise, as you mentioned, I would do that check in user-space). I think the loop I posted above handles that correctly, unless there is a bug again.

And I do think it's a bug to show extra "...", if there are no more arguments. It's super misleading.

> Again, this code has same bug, cause it just checks that i == max_args which could be done in userspace (which is done in this PR) and would have same result. As I wrote before:

I don't think so. You'll get to this comparison only if `argp != NULL`, which means there was supposed to be another argument, but we are not going to read it, because we are beyond our limit of arguments to read. Say, if you wanted 5 arguments and program had 5 arguments. Algorithm would go on 6th iteration, would see argp == NULL and would correctly conclude that there is no 6th argument. If it had 6+ arguments, we'd read argp, it would not be NULL, we wouldn't attempt to read its contents, but we'd know that there is 6+ arguments (that's where we do negation, extra bit, or flag, whatever). Makes sense? Did I miss anything here?

>> (btw, NULL initialization of argp is not necessary)
> Can you explain why ?

yes, because bpf_probe_read() guarantees that destination (i.e., argp) will be either NULL on error, or will contain correct value (which still can be NULL, but code ignores bpf_probe_read failures, so we'd assume there was no argument, which also sucks, but that's for another time).

Reply:
Regarding misleading behavior, currently if there is an argument that is bigger than `ARGSIZE -1` (which is `127`) it will be truncated and there will no any indication about that, which IMO is also misleading. For instance:

    /bin/echo abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789   # len(arg1) == 186
    # result from original execsnoop
    echo             25603  7487     0 /bin/echo abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz
ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcd

> yes, because bpf_probe_read() guarantees that destination (i.e., argp) will be either NULL on error, or will contain correct value

Thanks, I didn't know that, in that case I'll remove NULL and update PR shortly.